File: backend/src/models/fragment.ts
import { z } from "zod";

export const FragmentSourceType = z.enum([
  "note",
  "highlight",
  "bookmark",
  "article",
  "image",
  "file",
  "share",
]);
export type FragmentSourceType = z.infer<typeof FragmentSourceType>;

export const CreateFragmentSchema = z.object({
  title: z.string().min(1).max(500),
  content: z.string().min(1),
  sourceUrl: z.string().url().optional(),
  sourceType: FragmentSourceType.default("note"),
  metadata: z.record(z.unknown()).optional(),
  collectionId: z.string().uuid().optional(),
});

export const UpdateFragmentSchema = CreateFragmentSchema.partial();

export interface Fragment {
  id: string;
  userId: string;
  title: string;
  content: string;
  sourceUrl: string | null;
  sourceType: FragmentSourceType;
  embedding: number[] | null;
  metadata: Record<string, unknown>;
  createdAt: Date;
  updatedAt: Date;
}

export interface Connection {
  id: string;
  userId: string;
  fragmentAId: string;
  fragmentBId: string;
  similarity: number;
  aiSummary: string | null;
  createdAt: Date;
}

export interface Insight {
  id: string;
  userId: string;
  fragmentId: string | null;
  insightType: string;
  content: string;
  metadata: Record<string, unknown>;
  createdAt: Date;
}

File: apple/Shared/Models/Fragment.swift
import Foundation
import SwiftData

/// A single piece of captured knowledge — a note, highlight, bookmark, or clipping.
@Model
final class Fragment {
    @Attribute(.unique) var id: UUID
    var title: String
    var content: String
    var sourceURL: URL?
    var sourceType: SourceType
    var metadata: [String: String]
    var createdAt: Date
    var updatedAt: Date

    @Relationship(inverse: \Collection.fragments)
    var collections: [Collection]

    init(
        id: UUID = UUID(),
        title: String,
        content: String,
        sourceURL: URL? = nil,
        sourceType: SourceType = .note,
        metadata: [String: String] = [:],
        createdAt: Date = .now,
        updatedAt: Date = .now,
        collections: [Collection] = []
    ) {
        self.id = id
        self.title = title
        self.content = content
        self.sourceURL = sourceURL
        self.sourceType = sourceType
        self.metadata = metadata
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.collections = collections
    }

    enum SourceType: String, Codable, CaseIterable {
        case note
        case highlight
        case bookmark
        case article
        case image
        case file
        case share
    }
}
